watertap/unit_models/cstr_injection.py

```diff
@@ -0,0 +1,338 @@
+**default** - None.
+**Valid values:** {
+see reaction package for documentation.}""",
+        ),
+    )
+
+    def build(self):
+        """
+        Begin building model (pre-DAE transformation).
+        Args:
+            None
+        Returns:
+            None
+        """
+        # Call UnitModel.build to setup dynamics
+        super(CSTR_InjectionData, self).build()
+
+        # Build Control Volume
+        self.control_volume = ControlVolume0DBlock(
+            default={
+                "dynamic": self.config.dynamic,
+                "has_holdup": self.config.has_holdup,
+                "property_package": self.config.property_package,
+                "property_package_args": self.config.property_package_args,
+                "reaction_package": self.config.reaction_package,
+                "reaction_package_args": self.config.reaction_package_args,
+            }
+        )
+
+        self.control_volume.add_geometry()
+
+        self.control_volume.add_state_blocks(
+            has_phase_equilibrium=self.config.has_phase_equilibrium
+        )
+
+        self.control_volume.add_reaction_blocks(
+            has_equilibrium=self.config.has_equilibrium_reactions
+        )
+
+        self.control_volume.add_material_balances(
+            balance_type=self.config.material_balance_type,
+            has_rate_reactions=True,
+            has_equilibrium_reactions=self.config.has_equilibrium_reactions,
+            has_phase_equilibrium=self.config.has_phase_equilibrium,
+            has_mass_transfer=True,
+        )
+
+        self.control_volume.add_energy_balances(
+            balance_type=self.config.energy_balance_type,
+            has_heat_of_reaction=self.config.has_heat_of_reaction,
+            has_heat_transfer=self.config.has_heat_transfer,
+        )
+
+        self.control_volume.add_momentum_balances(
+            balance_type=self.config.momentum_balance_type,
+            has_pressure_change=self.config.has_pressure_change,
+        )
+
+        # Add Ports
+        self.add_inlet_port()
+        self.add_outlet_port()
+
+        # Add object references
+        self.volume = Reference(self.control_volume.volume[:])
+        self.injection = Reference(self.control_volume.mass_transfer_term[...])
+
+        # Add CSTR performance equation
+        @self.Constraint(
+            self.flowsheet().time,
+            self.config.reaction_package.rate_reaction_idx,
+            doc="CSTR performance equation",
+        )
+        def cstr_performance_eqn(b, t, r):
+            return b.control_volume.rate_reaction_extent[t, r] == (
+                b.volume[t] * b.control_volume.reactions[t].reaction_rate[r]
+            )
+
+        # Set references to balance terms at unit level
+        if (
+            self.config.has_heat_transfer is True
+            and self.config.energy_balance_type != EnergyBalanceType.none
+        ):
+            self.heat_duty = Reference(self.control_volume.heat[:])
```

Comment:
Still not 100% clear on when/why we'd choose to use `Reference` over `add_object_reference`

Reply:
In general you should use `Reference` wherever possible.

There are two exceptions where using `add_object_reference` is appropriate:

1. When you want to reference a scalar object - using `Reference` will make it appear to be indexed by `[None]` which is unfortunate and likely confusing (there are reasons behind this).
2. If you explicitly do not want the `Reference` to appear in `component_objects`. `References` get picked up by this, so when iterating over all components the referenced component can show up twice. There are other ways to avoid this however, so this should only be done where there is a very good cause.